### PR TITLE
Initial eigen port of DecodingReturnValues

### PR DESCRIPTION
**Pull request by abhidg**
This commit ports sumOverPairs{,00,01,11} to use Eigen Arrays. Eigen
arrays are automatically converted to numpy arrays by pybind and support
useful array and matrix operations.

### CODE REVIEW

File: ASMC_SRC/SRC/pybind.cpp [hunk 1]
```diff
@@ -55,7 +56,7 @@ PYBIND11_MODULE(pyASMC, m) {
     py::bind_vector<std::vector<Individual>>(m, "VectorIndividual");
     py::bind_vector<std::vector<uint>>(m, "VectorUInt");
     py::bind_vector<std::vector<PairObservations>>(m, "VectorPairObservations");
-    py::bind_vector<std::vector<std::vector<float>>>(m, "Matrix");
+    //py::bind_vector<std::vector<std::vector<float>>>(m, "Matrix");
```
Comment (brianzhang01) [inline ASMC_SRC/SRC/pybind.cpp]: Keep in mind that this change will also affect the wrapping of DecodingQuantities::classicEmissionTable and compressedEmissionTable which are of type vector<vector<float>>. As we haven't figured out what we want to do in those cases, I think we should leave the Matrix type in there?

**Comment by abhidg** [comment]
>     * Have you done this before, and how detailed of a look have you had of the [docs](https://pybind11.readthedocs.io/en/master/advanced/cast/eigen.html)? I guess the docs focus on passing arguments / returning, and not as much on our case of wrapping attributes.

This is my first time porting to Eigen! It seems user-friendly. Pybind should automatically convert this to a numpy array. I will write #22 this afternoon, and if it passes, we can merge this.

>     * I'm more used to Eigen::MatrixXf instead of Eigen::ArrayXXf. Do you have thoughts on why we would use one over the other?

From the [docs](https://eigen.tuxfamily.org/dox/group__TutorialArrayClass.html), Matrix is suited for linear algebra operations, which we are not using. Array supports coefficient-wise operations. For the purposes of `sumOverPairs`, I think they are pretty much interchangeable.

**Comment by brianzhang01** [comment]
Actually, I think we need a Python test. Right now there's nothing that tries to build the ASMC HMM from Python and then access sumOverPairs. Without that, there's no way to know if the Python wrapping is actually working (unless you've tested this on your own?).

This doesn't need to be a regression test -- let's just build the HMM (which should zero everything in DecodingReturnValues), and check from Python that we can access the sumOverPairs{, 00, 01, 11}, that the values are 0, and that the shapes are correct. It can be modeled off of `test_regression.cpp`.

**Comment by brianzhang01** [comment]
Great, thanks. I'm more used to MatrixXf but yeah I saw that as well in the docs re coefficient operations. If you look at `HMM::elementWiseMultVectorScalar`, `HMM::elementWiseMultVectorVector`, `HMM::elementWiseMultMatrixMatrix`, these are things we might end up porting to Eigen later, and which would be simplified if we go with the Array classes.

Just to clarify, I think it would be nice if we have both a Python unit test, which doesn't decode anything and just checks the shapes of the sumOverPairs, and a separate regression test (#22) which does decode. If you write the unit test first and put that in this pull request, we can merge that first, and do the Python regression test later.
